fv3config/_asset_list.py

```diff
@@ -102,71 +116,45 @@ def get_asset_dict(source_location, source_name, target_location='',
-        target_location (str, optional): target_location used for generated assets.
-            Defaults to '' which is root of run-directory.
-        copy_method (str, optional): copy_method used for generated assets. Defaults
-            to 'copy'.
-
-    Returns:
-        list: a list of asset dictionaries
-    """
-    asset_list = []
-    for root, dirs, files in os.walk(source_directory):
-        if root == source_directory:
-            file_target_location = target_location
-        else:
-            file_target_location = os.path.join(target_location,
-                                                os.path.basename(root))
-        for file in files:
-            asset_list.append(get_asset_dict(os.path.join(source_directory, root),
-                                             file,
-                                             target_location=file_target_location,
-                                             copy_method=copy_method))
-    return asset_list
+def _get_gcloud_fs():
+    return gcsfs.GCSFileSystem(project=_get_gcloud_project())
 
 
-def asset_list_from_gs_bucket(url, target_location=''):
-    """Return asset_list from all files that begin with a google cloud storage url.
-    Will not recurse to files that are in "sub-directories" of the url.
+def asset_list_from_path(location, target_location='', copy_method='copy'):
+    """Return asset_list from all files within a given path.
 
     Args:
-        url (str): google cloud storage url from which to generate asset_list.
+        location (str): local path or google cloud storage url.
         target_location (str, optional): target_location used for generated assets.
             Defaults to '' which is root of run-directory.
+        copy_method ('copy' or 'link', optional): whether to copy or link assets,
+            defaults to 'copy'. If location is a google cloud storage url, this option
+            is ignored and files are copied.
 
     Returns:
         list: a list of asset dictionaries
         """
+    if is_gsbucket_url(location):
+        copy_method = 'copy'
+        fs = _get_gcloud_fs()
+    else:
+        fs = LOCAL_FS
     asset_list = []
-    # TODO: use gcsfs instead of gsutil for following code
-    stdout_str = Popen(['gsutil', 'ls', url], stdout=PIPE).stdout.read()
-    path_list = stdout_str.decode().split('\n')[:-1]
-    for path in path_list:
-        dirname, basename = os.path.split(path)
-        if basename != '':
-            asset_list.append(get_asset_dict(dirname,
-                                             basename,
-                                             target_location=target_location))
+    path_list = fs.walk(location)
+    for dirname, _, files in path_list:
+        print(dirname, files, target_location)
+        subdir_target_location = os.path.join(
+            target_location, os.path.relpath(dirname, start=location)
+        )
+        for basename in files:
+            asset_list.append(
```

Comment:
In `gcsfs`, there is no `gs://` prefix. Are you putting them back in anywhere? If not, then I do not see how the generated `asset` will not contain any information about whether the file is remote or local. 

However, instead of re-prepending the `gs://` to the path, I would recommend adding a `protocol` entry to the asset, which can be used to instantiate the filesystem in `copy_file`. To write a file this would look like:
```
def copy_file(protocol, source_path, target_path):
    fs = fsspec.filesystems(protocol)
    fs.get(source_path, target_path)
```
This would change the asset API, but so far I don't think any of us are using that directly.

Reply:
Yes, I ran into that bug when working on the fv3run kubernetes PR and prepended it there. Part of why I didn't ask for reviews on this PR yet was that the gcs functionality couldn't be tested until the kubernetes stuff was working, and now we have that as an integration test (and a hint as to at least one useful unit test I can add).

The main reason I'd opt for pre-pending gs:// instead of adding `protocol` is it makes the two APIs (location and asset) consistent with one another. Because the asset API *is* a user-facing api (even if none of us are using it currently), I think consistency of the APIs is worthwhile. It's also not hard to enforce or make simple to access/generalize, we could write a `get_protocol(path)` function (I effectively did this, but `get_fs(path)`, in the upcoming kubernetes API).

Looking at the changes in the kubernetes branch, the code has actually been changed a fair bit more -- I think it may be best to review them as a single PR instead of working to pull just the filesystem changes out.

Reply:
It also ends up being a little more complicated than picking one protocol, because there are separate `put`, `get`, and `copy` methods for copying to, from, or within a filesystem. The code I wrote in the next PR handles these different cases.